staker-rewards/lib/fraud-detector.js

```diff
@@ -0,0 +1,110 @@
+import clc from "cli-color"
+
+export default class FraudDetector {
+  constructor(
+    context,
+    tokenStaking,
+    bondedECDSAKeepFactory,
+    tbtcSystem,
+    factoryDeploymentBlock
+  ) {
+    this.context = context
+
+    this.tokenStaking = tokenStaking
+    this.bondedECDSAKeepFactory = bondedECDSAKeepFactory
+    this.tbtcSystem = tbtcSystem
+
+    this.factoryDeploymentBlock = factoryDeploymentBlock
+  }
+
+  static async initialize(context) {
+    const {
+      TokenStaking,
+      BondedECDSAKeepFactory,
+      TBTCSystem,
+      factoryDeploymentBlock,
+    } = context.contracts
+
+    const tokenStaking = await TokenStaking.deployed()
+    const bondedECDSAKeepFactory = await BondedECDSAKeepFactory.deployed()
+    const tbtcSystem = await TBTCSystem.deployed()
+
+    return new FraudDetector(
+      context,
+      tokenStaking,
+      bondedECDSAKeepFactory,
+      tbtcSystem,
+      factoryDeploymentBlock
+    )
+  }
+
+  async isOperatorKeepECDSAFraudulent(operator) {
+    // Get all slashing events for the operator.
+    const events = await this.tokenStaking.getPastEvents("TokensSlashed", {
+      fromBlock: this.factoryDeploymentBlock,
+      toBlock: "latest",
+      filter: { operator: operator },
+    })
+
+    if (events.length > 0) {
+      console.log(
+        clc.yellow(
+          `found slashing event for operator [${operator}];` +
+            ` please double check correctness of a calculation result`
+        )
+      )
+    } else {
+      console.debug(
+        `found [${events.length}] slashing events for operator [${operator}]`
+      )
+    }
```

Comment:
It is obviously a late discovery but worth considering if those lines are not all we need. As we discussed yesterday:
- In ECDSA keeps, we slash KEEP tokens and seize ETH bonds,
- In the random beacon, we seize KEEP tokens.

Also, the only situation when we slash KEEP tokens in ECDSA keeps is a signature fraud.

Removing the rest of the validation will not only simplify the code but will also decouple staker rewards implementation from tBTC. Other advantages I see:
- Authorizing another application for staking rewards is simple: we add another check regarding the existence of the operator in a sortition pool and make sure signature-fraud-level misbehavior on the application side leads to slashing tokens. This is important because the only place in the code where staker rewards will be aware of tBTC is the sortition pool check. It is quite elegant, especially that the sortition pool is a Keep concept, not an application concept.
- We can test this logic without a lot of mocking, see [here](https://github.com/keep-network/keep-ecdsa/blob/329533d8f877851f0a9a2bf0a21b7032baec928a/staker-rewards/test/sla-calculator.test.js#L19-L37) please.
- Rules are simple: you seriously misbehave, your tokens are slashed - no rewards.

I was thinking whether the validation below does not make it safer but I do not think it does. We do know the only scenario tokens are slashed is the signature fraud and in the unlikely case we sanction another application for staking rewards and that application seizes tokens in case of serious misbehavior, we'll have to implement similar logic as below anyway but only for that application.

Reply:
I had similiar thoughts yesterday after our discussion. I'm glad you brought this up. I'll simplify the script.